Add "undo_max_stack_size" property to TextEdit

CaptainProton42 opened the pull request:
Previously, there was no memory limit on TextEdit's undo stack. This PR adds "undo_max_stack_size" as a property to TextEdit to prevent potential memory leaks.

This fixes #37838.

CaptainProton42 committed:
Limit undo stack size

The stack size of the undo history of a TextEdit was not limited leading
to potential memory leaks when doing lots of operations on a TextEdit.

This commit adds the option gui/common/text_edit_undo_stack_max_size
to the project settings. The first element of the undo stack is popped
if the stack's size exceeds this value ensuring limited memory usage.

The default stack size setting is 1024.

Fixes #37838.

Calinou commented:
This is a good idea, but can't we hardcore a sensible default value instead (or expose it in the Project Settings)?

CaptainProton42 commented:
I added 'gui/common/text_edit_undo_stack_max_size' to the project settings and removed the node property from the editor. 'undo_stack_max_size' now gets its value from the global in the constructor. I chose a size of 128 as a default, I think 128 undo operations should be plenty.

groud commented:
>  I chose a size of 128 as a default, I think 128 undo operations should be plenty.

I would go for a significantly higher number. Computers have a lot of memory nowadays, so it's better to increase this size as much as possible. I would go for at least 1028.

CaptainProton42 commented:
Fair enough, I bumped it up to 1024.

Paulb23 reviewed (approved):
LGTM

akien-mga commented:
Could you squash commits into one?

CaptainProton42 commented:
Done! (Also learned a cool new git trick, thanks :P)

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.2.2.
